File: packages/sdk-flutter/lib/src/models/exceptions.dart
import 'package:meta/meta.dart';

/// Base exception for all ISL SDK errors
@immutable
sealed class ISLException implements Exception {
  final String message;
  final String? code;
  final Map<String, dynamic>? details;
  final StackTrace? stackTrace;

  const ISLException({
    required this.message,
    this.code,
    this.details,
    this.stackTrace,
  });

  @override
  String toString() => 'ISLException($code): $message';
}

/// Thrown when input validation fails (precondition violation)
@immutable
class ValidationException extends ISLException {
  final String field;
  final dynamic invalidValue;
  final List<String> constraints;

  const ValidationException({
    required super.message,
    required this.field,
    this.invalidValue,
    this.constraints = const [],
    super.code = 'VALIDATION_ERROR',
    super.details,
    super.stackTrace,
  });

  factory ValidationException.required(String field) => ValidationException(
        message: '$field is required',
        field: field,
        constraints: ['required'],
      );

  factory ValidationException.invalidFormat(String field, String expected) =>
      ValidationException(
        message: '$field has invalid format. Expected: $expected',
        field: field,
        constraints: ['format:$expected'],
      );

  factory ValidationException.outOfRange(
    String field, {
    num? min,
    num? max,
    num? actual,
  }) =>
      ValidationException(
        message: '$field must be between $min and $max',
        field: field,
        invalidValue: actual,
        constraints: [
          if (min != null) 'min:$min',
          if (max != null) 'max:$max',
        ],
      );

  @override
  String toString() => 'ValidationException($field): $message';
}

/// Thrown when a postcondition verification fails
@immutable
class ContractViolationException extends ISLException {
  final String contractType;
  final String expected;
  final String actual;

  const ContractViolationException({
    required super.message,
    required this.contractType,
    required this.expected,
    required this.actual,
    super.code = 'CONTRACT_VIOLATION',
    super.details,
    super.stackTrace,
  });

  factory ContractViolationException.postcondition({
    required String condition,
    required String expected,
    required String actual,
  }) =>
      ContractViolationException(
        message: 'Postcondition failed: $condition',
        contractType: 'postcondition',
        expected: expected,
        actual: actual,
      );

  factory ContractViolationException.invariant({
    required String condition,
    required String expected,
    required String actual,
  }) =>
      ContractViolationException(
        message: 'Invariant violated: $condition',
        contractType: 'invariant',
        expected: expected,
        actual: actual,
      );

  @override
  String toString() =>
      'ContractViolationException($contractType): expected $expected, got $actual';
}

/// Thrown when network operations fail
@immutable
class NetworkException extends ISLException {
  final int? statusCode;
  final String? url;
  final Duration? timeout;

  const NetworkException({
    required super.message,
    this.statusCode,
    this.url,
    this.timeout,
    super.code = 'NETWORK_ERROR',
    super.details,
    super.stackTrace,
  });

  factory NetworkException.timeout(String url, Duration timeout) =>
      NetworkException(
        message: 'Request timed out after ${timeout.inSeconds}s',
        url: url,
        timeout: timeout,
        code: 'TIMEOUT',
      );

  factory NetworkException.noConnection() => const NetworkException(
        message: 'No internet connection',
        code: 'NO_CONNECTION',
      );

  factory NetworkException.serverError(int statusCode, String? body) =>
      NetworkException(
        message: 'Server error: $statusCode',
        statusCode: statusCode,
        code: 'SERVER_ERROR',
        details: body != null ? {'body': body} : null,
      );

  bool get isTimeout => code == 'TIMEOUT';
  bool get isNoConnection => code == 'NO_CONNECTION';
  bool get isServerError => statusCode != null && statusCode! >= 500;

  @override
  String toString() => 'NetworkException($code): $message';
}

/// Thrown when authentication fails
@immutable
class AuthenticationException extends ISLException {
  final AuthFailureReason reason;

  const AuthenticationException({
    required super.message,
    required this.reason,
    super.code = 'AUTH_ERROR',
    super.details,
    super.stackTrace,
  });

  factory AuthenticationException.invalidCredentials() =>
      const AuthenticationException(
        message: 'Invalid credentials',
        reason: AuthFailureReason.invalidCredentials,
      );

  factory AuthenticationException.tokenExpired() =>
      const AuthenticationException(
        message: 'Authentication token has expired',
        reason: AuthFailureReason.tokenExpired,
      );

  factory AuthenticationException.refreshFailed() =>
      const AuthenticationException(
        message: 'Failed to refresh authentication token',
        reason: AuthFailureReason.refreshFailed,
      );

  factory AuthenticationException.unauthorized() =>
      const AuthenticationException(
        message: 'Unauthorized access',
        reason: AuthFailureReason.unauthorized,
      );

  @override
  String toString() => 'AuthenticationException($reason): $message';
}

enum AuthFailureReason {
  invalidCredentials,
  tokenExpired,
  refreshFailed,
  unauthorized,
  mfaRequired,
  accountLocked,
}

/// Thrown when rate limiting is triggered
@immutable
class RateLimitException extends ISLException {
  final Duration retryAfter;
  final int? limit;
  final int? remaining;
  final DateTime? resetAt;

  const RateLimitException({
    required super.message,
    required this.retryAfter,
    this.limit,
    this.remaining,
    this.resetAt,
    super.code = 'RATE_LIMITED',
    super.details,
    super.stackTrace,
  });

  factory RateLimitException.fromHeaders(Map<String, dynamic> headers) {
    final retryAfterSecs =
        int.tryParse(headers['Retry-After']?.toString() ?? '60') ?? 60;
    final limit = int.tryParse(headers['X-RateLimit-Limit']?.toString() ?? '');
    final remaining =
        int.tryParse(headers['X-RateLimit-Remaining']?.toString() ?? '');
    final resetTimestamp =
        int.tryParse(headers['X-RateLimit-Reset']?.toString() ?? '');

    return RateLimitException(
      message: 'Rate limit exceeded. Retry after ${retryAfterSecs}s',
      retryAfter: Duration(seconds: retryAfterSecs),
      limit: limit,
      remaining: remaining,
      resetAt: resetTimestamp != null
          ? DateTime.fromMillisecondsSinceEpoch(resetTimestamp * 1000)
          : null,
    );
  }

  @override
  String toString() =>
      'RateLimitException: retry after ${retryAfter.inSeconds}s';
}

File: packages/sdk-python/src/isl/exceptions.py
"""
ISL Exceptions - Exception hierarchy for ISL SDK.

This module provides a hierarchy of exceptions for different error types.
"""

from __future__ import annotations

from typing import Any


class ISLError(Exception):
    """Base exception for all ISL SDK errors."""
    
    def __init__(self, message: str) -> None:
        self.message = message
        super().__init__(message)


class ValidationError(ISLError):
    """Validation error for input data."""
    
    def __init__(
        self,
        message: str,
        field: str | None = None,
        value: Any = None,
    ) -> None:
        self.field = field
        self.value = value
        super().__init__(message)


class PreconditionError(ISLError):
    """
    Precondition violation error.
    
    Raised when ISL preconditions are not met before an API call.
    """
    
    def __init__(
        self,
        message: str,
        precondition: str,
        actual_value: Any = None,
    ) -> None:
        self.precondition = precondition
        self.actual_value = actual_value
        super().__init__(f"Precondition violated: {message}")


class PostconditionError(ISLError):
    """
    Postcondition violation error.
    
    Raised when server response doesn't match ISL contract.
    """
    
    def __init__(
        self,
        message: str,
        postcondition: str,
        expected_value: Any = None,
        actual_value: Any = None,
    ) -> None:
        self.postcondition = postcondition
        self.expected_value = expected_value
        self.actual_value = actual_value
        super().__init__(
            f"Postcondition violated: {message} "
            f"(expected: {expected_value}, actual: {actual_value})"
        )


class NetworkError(ISLError):
    """
    Network error.
    
    Raised for connection failures, timeouts, etc.
    """
    
    def __init__(
        self,
        message: str,
        cause: Exception | None = None,
        is_timeout: bool = False,
        is_connection_error: bool = False,
    ) -> None:
        self.cause = cause
        self.is_timeout = is_timeout
        self.is_connection_error = is_connection_error
        super().__init__(f"Network error: {message}")


class ServerError(ISLError):
    """
    Server error.
    
    Raised for 5xx responses.
    """
    
    def __init__(
        self,
        message: str,
        status_code: int | None = None,
        error_code: str | None = None,
    ) -> None:
        self.status_code = status_code
        self.error_code = error_code
        super().__init__(f"Server error: {message}")


class UnauthorizedError(ISLError):
    """Unauthorized error (401)."""
    
    def __init__(self, message: str = "Unauthorized") -> None:
        super().__init__(message)


class ForbiddenError(ISLError):
    """Forbidden error (403)."""
    
    def __init__(self, message: str = "Forbidden") -> None:
        super().__init__(message)


class NotFoundError(ISLError):
    """Not found error (404)."""
    
    def __init__(
        self,
        message: str = "Resource not found",
        resource_type: str | None = None,
        resource_id: str | None = None,
    ) -> None:
        self.resource_type = resource_type
        self.resource_id = resource_id
        super().__init__(message)


class ConflictError(ISLError):
    """Conflict error (409)."""
    
    def __init__(
        self,
        message: str,
        conflict_field: str | None = None,
    ) -> None:
        self.conflict_field = conflict_field
        super().__init__(message)


class RateLimitError(ISLError):
    """Rate limit error (429)."""
    
    def __init__(
        self,
        retry_after_seconds: int,
        message: str = "Rate limited",
    ) -> None:
        self.retry_after_seconds = retry_after_seconds
        super().__init__(f"{message}. Retry after {retry_after_seconds} seconds")


class TimeoutError(ISLError):
    """Request timeout error."""
    
    def __init__(
        self,
        message: str = "Request timed out",
        timeout_ms: int | None = None,
    ) -> None:
        self.timeout_ms = timeout_ms
        super().__init__(message)


class SerializationError(ISLError):
    """Serialization/deserialization error."""
    
    def __init__(
        self,
        message: str,
        cause: Exception | None = None,
        response_body: str | None = None,
    ) -> None:
        self.cause = cause
        self.response_body = response_body
        super().__init__(f"Serialization error: {message}")
